### ISSUE TITLE
Any way to get hidden states out of rolling_batch=vllm?

### ISSUE THREAD

**Issue by ramkrishs** [deepjavalibrary/djl-serving#3066]
# Feature request draft: djl-serving

Target repo: `deepjavalibrary/djl-serving`

---

## Title

Any way to get hidden states out of `rolling_batch=vllm`?

## The problem

We serve a small fine-tuned Llama model (3B, LoRA) with `option.rolling_batch=vllm`, and on top
of the normal generated text we want to return a confidence score for each response. We compute it
with a tiny linear probe (a few thousand parameters) run on the last hidden state of the final
generated token — basically free, since that hidden state already exists as part of the forward
pass, we just need a way to read it out.

The only way we've found to do this is to drop `rolling_batch=vllm` entirely and write a custom
`model.py` that calls `transformers.generate(..., output_hidden_states=True)` directly. That works,
but it means giving up vLLM's batching. We benchmarked both on the same GPU (A10G): the custom
handler does about 30 tok/s no matter how many requests come in at once (no batching), while
`rolling_batch=vllm` does 500-1000+ tok/s at batch sizes 16-128. So getting the confidence score
currently costs us somewhere around 15-30x throughput.

We tried the documented pattern for customizing rolling-batch output (`HuggingFaceService` +
`@output_formatter` in `model.py`), hoping we could keep vLLM batching and just add our own
post-processing on the response. But looking through `request_io.py` and `utils.py`, it looks like
hidden states just aren't kept anywhere in that path — `Token`/`Sequence`/`TextGenerationOutput`
only carry token ids, decoded text, and log-probs. So there's nothing for an output formatter to
read, no matter how it's written.

## What we'd want

Some opt-in way to keep the last token's hidden state around when using `rolling_batch=vllm`, so a
custom `@output_formatter` could read it off the result and do something with it (a confidence
probe like ours, or anything else people might want a hidden state for).

Rough shape of what that could look like:

- A `serving.properties` flag, something like `option.output_last_hidden_state=true`, off by
  default since it's extra memory most people won't want.
- When it's on, `TextGenerationOutput`/`Sequence` gets an extra optional field with the hidden
  state for the last token of that sequence (not the whole sequence — just the last token, to keep
  the cost bounded).
- Then `@output_formatter` code can just read it off the result the same way it already reads
  `.finish_reason` or the token list.

## Why we haven't just done this ourselves

We think the real limitation is actually one level down, in vLLM itself — its normal generation
path doesn't keep hidden states around once it's done computing logits (there's a separate
embeddings/pooling API that does, but that's a different mode, not regular generation). If that's
right, this isn't really fixable purely on the djl-serving side without vLLM exposing that data
first. Would appreciate a sanity check from maintainers here — if this is fundamentally a vLLM
question, we're happy to raise it there instead and treat this issue as a follow-up once that
exists.

## What we're doing in the meantime

- Running the custom `model.py` + plain HF generate path for now, accepting the throughput hit,
  since we need the confidence score more than we need max throughput on this particular traffic.
- Considered running two separate deployments (one vLLM for bulk traffic, one custom-handler for
  a smaller "scored" slice) — works, but it's more infra to run and needs routing logic on our end.
- Considered a separate call to a judge/classifier model instead of a probe - defeats the point,
  since the whole appeal here was doing this at effectively zero extra cost.

## Environment

- `djl-inference:0.36.0-lmi26.0.0-cu130-v1.0`
- Llama-3.2-3B-Instruct + LoRA, merged fp16, single A10G

Happy to share our benchmark numbers or a small repro if that's useful.

**Comment by ramkrishs** [comment]
## Update: found a working mechanism, benchmarked the cost

Wanted to follow up on this with something concrete rather than leave it as a pure feature ask.

Turns out vLLM already has a real mechanism for this — `extract_hidden_states`
(merged into vLLM main back in March, via a KV-connector + dummy-Eagle3-draft-model
trick, not a simple flag). It genuinely works: we ran it against our own model
(Llama-3.2-3B-Instruct + LoRA, 3B params) on a single A10G and got real,
correctly-shaped hidden states (`[num_tokens, 1, 3072]`) back per request via
`kv_transfer_params.hidden_states_path`, using:

```
--no-enable-chunked-prefill
--speculative_config '{"method": "extract_hidden_states", "num_speculative_tokens": 1, "draft_model_config": {"hf_config": {"eagle_aux_hidden_state_layer_ids": [28]}}}'
--kv_transfer_config '{"kv_connector": "ExampleHiddenStatesConnector", "kv_role": "kv_producer", "kv_connector_extra_config": {"shared_storage_path": "..."}}'
```

So the actual gap isn't in vLLM anymore — it's that DJL's `VLLMRollingBatch` wrapper
doesn't read `kv_transfer_params` off the output at all today, so none of this is
reachable from a DJL `@output_formatter`. That reframes the ask: this issue is really
about wiring DJL up to consume something vLLM already exposes, not about getting vLLM
to expose something new.

### The catch: it's not free, and the cost isn't flat

The docs note chunked prefill has to be disabled to use this, so we benchmarked the
actual throughput cost on our GPU/model, at three concurrency levels, baseline
(chunked prefill on, no extraction) vs. extract_hidden_states (chunked prefill off):

| Batch size | Baseline tok/s | Extract tok/s | Throughput cost | Baseline p99 | Extract p99 |
|---|---|---|---|---|---|
| 32  | 1322.2 | 1179.3 | -10.8% | 2.41s | 2.69s |
| 64  | 2005.6 | 1543.0 | -23.1% | 3.17s | 3.89s |
| 128 | 2486.2 | 1724.4 | -30.6% | 5.07s | 6.95s |

All runs: 0 failures, same 40-prompt pool, 60s each.

The cost isn't a flat tax — it roughly triples from bs32 to bs128 and hadn't
plateaued by bs128. Makes sense mechanistically: without chunked prefill, large
prefill batches block decode for in-flight requests instead of interleaving with
it, so the penalty compounds as concurrency rises.

### What we think is worth doing here

Given the above, we'd frame the real ask for this issue as: an opt-in path in
`VLLMRollingBatch`/`djl_python` to read `kv_transfer_params.hidden_states_path`
(when a request sets it) and thread the loaded hidden state into
`Sequence`/`TextGenerationOutput` so a custom `@output_formatter` can read it —
essentially "make the already-working vLLM mechanism reachable from DJL," clearly
documented with the throughput tradeoff above so people can make an informed call
based on their own concurrency needs, rather than this looking like a free feature.

Happy to have a stab at that if it's a direction maintainers are open to — figured
real numbers were more useful than a bigger ask before that conversation.

**Comment by ramkrishs** [comment]
Opened a PR for this: #3067 — wires up hidden_states_path so it's actually readable through VLLMRollingBatch, plus the throughput numbers from the last comment.

**Comment by ramkrishs** [comment]
Quick correction to my earlier comment — I'd said custom output formatters were broken for rolling_batch=vllm, but that's only the legacy sync path. Tested the async path (the default since LMI v17) directly and it already reads hidden_states_path fine through a custom formatter, no changes needed. Nothing to fix there.
